src/main/sensors/adcinternal.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#if defined(USE_ADC) && defined(USE_ADC_INTERNAL)
+
+#include "build/debug.h"
+
+#include "common/utils.h"
+
+#include "drivers/adc.h"
+
+typedef struct movingAverageStateUint16_s {
+    uint32_t sum;
+    uint16_t *values;
+    uint8_t size;
+    uint8_t pos;
+} movingAverageStateUint16_t;
+
+uint16_t updateMovingAverageUint16(movingAverageStateUint16_t *state, uint16_t newValue)
+{
+    state->sum -= state->values[state->pos];
+    state->values[state->pos] = newValue;
+    state->sum += newValue;
+    state->pos = (state->pos + 1) % state->size;
+
+    return state->sum / state->size;
+}
+
+static uint16_t adcVrefintValue;
+static uint16_t adcVrefintValues[8];
+movingAverageStateUint16_t adcVrefintAverageState = { 0, adcVrefintValues, 8, 0 } ;
+
+static uint16_t adcTempsensorValue;
+static uint16_t adcTempsensorValues[8];
+movingAverageStateUint16_t adcTempsensorAverageState = { 0, adcTempsensorValues, 8, 0 } ;
+
+static int16_t coreTemperature;
+
+uint16_t getVrefMv(void)
+{
+#ifdef ADC_VOLTAGE_REFERENCE_MV
+    return ADC_VOLTAGE_REFERENCE_MV;
+#else
+    return 3300 * adcVrefintValue / adcVREFINTCAL;
+#endif
+}
+
+uint16_t getCoreTemperatureCelsius(void)
+{
+    return coreTemperature;
+}
+
+void adcInternalProcess(timeUs_t currentTimeUs)
+{
+    UNUSED(currentTimeUs);
+
+    if (adcInternalIsBusy()) {
+        return;
+    }
+
+    uint16_t vrefintSample = adcInternalReadVrefint();
+    uint16_t tempsensorSample = adcInternalReadTempsensor();
+
+    adcVrefintValue = updateMovingAverageUint16(&adcVrefintAverageState, vrefintSample);
+    adcTempsensorValue = updateMovingAverageUint16(&adcTempsensorAverageState, tempsensorSample);
+
+    int32_t adcTempsensorAdjusted = (int32_t)adcTempsensorValue * 3300 / getVrefMv();
```

Comment:
quick question VRefint on F4 and also other STM32 is set at 1.2V, you can check that in datasheet at 5.3.24, shouldnt this calculation be something like:
```int32_t adcTempsensorAdjusted = (int32_t)adcTempsensorValue * 1200 / getVrefMv();```
assuming ```getVrefMv()``` returns raw value read from adc?

Reply:
@conkerkh 
- I'm using VREFINTCAL, the ADC value of VREFINT at VREF=3.3V, instead of fixed 1.2V.
- `adcTempsensorAdjusted` calculation is a compromise. I could use VREFINTCAL here, too, but doing so will add extra `#ifdef ADC_VOLTAGE_REFERENCE_MV` conditional block.

Reply:
Ah so end result is the same very good:) I hope this will be merged soon I'm curious how hot is my MCU!